File: backend/tests/sse.py
# SPDX-License-Identifier: Apache-2.0
# Copyright The Robinauts Authors

"""Reading a stream the way a browser does: the SSE parser, and a driver.

Two things, and the second exists because of a limit in the first tool a test
would reach for. ``httpx``'s ASGI transport **runs the application to its end**
before it answers: every byte of the response is collected and handed over at
once. That is fine for a stream that finishes -- a turn whose engine the test
lets run to the end -- and it is no use at all for what a stream is *for*:
receiving one event while the run goes on, and going away in the middle.

So ``streaming`` drives the ASGI application itself, as a server does: it
sends the request, it hands each body chunk to the test as it is written, and
``close`` sends ``http.disconnect``, which is exactly what a server sends when
the client's socket goes. Nothing here knows anything about the routes; it is
the protocol, and about forty lines of it.

The parser is deliberately literal about the format: an ``id:``, an ``event:``
and the ``data:`` lines of one block, and a block that is only a comment
(``: keep-alive``) kept as one, because a heartbeat is a thing a test asks
about.
"""

from __future__ import annotations

import asyncio
import json
from collections.abc import AsyncIterator, Mapping
from contextlib import asynccontextmanager
from dataclasses import dataclass, field
from typing import Any

BLOCK = "\n\n"
"""What separates one server-sent event from the next."""


@dataclass(frozen=True, slots=True)
class Sent:
    """One block of a stream: an event, or a comment on its own."""

    data: str = ""
    name: str | None = None
    """The ``event:`` field, which is the AG-UI type here."""
    id: str | None = None
    """The ``id:`` field, which is the event's position in its run."""
    comment: str | None = None
    """What a block that carried nothing but a comment said."""

    @property
    def body(self) -> dict[str, Any]:
        """The event's JSON, as the client would read it."""
        read = json.loads(self.data)
        assert isinstance(read, dict)
        return read

    @property
    def type(self) -> str:
        """The AG-UI type the body names; what a test reads at a glance."""
        return str(self.body["type"])


def parse(text: str) -> list[Sent]:
    """Every block of that stream, in order, comments included."""
    blocks = []
    for block in text.split(BLOCK):
        if not block.strip("\n"):
            continue
        blocks.append(_block(block))
    return blocks


def events(text: str) -> list[Sent]:
    """The blocks that carry an event; the heartbeats left out."""
    return [block for block in parse(text) if block.comment is None]


def beats(text: str) -> list[str]:
    """What the comment blocks of that stream said."""
    return [block.comment for block in parse(text) if block.comment is not None]


def _block(block: str) -> Sent:
    """One block, field by field. Unknown fields are ignored, as a client does."""
    data: list[str] = []
    name: str | None = None
    identifier: str | None = None
    comment: str | None = None
    for line in block.split("\n"):
        if not line:
            continue
        if line.startswith(":"):
            comment = line[1:].strip()
            continue
        field_name, _, value = line.partition(":")
        value = value[1:] if value.startswith(" ") else value
        if field_name == "data":
            data.append(value)
        elif field_name == "event":
            name = value
        elif field_name == "id":
            identifier = value
    return Sent(
        data="\n".join(data),
        name=name,
        id=identifier,
        comment=comment if not data else None,
    )


@dataclass
class Streamed:
    """One request being served, and what it has written so far."""

    status: int
    headers: Mapping[str, str]
    _chunks: asyncio.Queue[bytes | None]
    _served: asyncio.Task[None]
    _gone: asyncio.Event
    _read: str = ""
    _over: bool = False
    _blocks: list[Sent] = field(default_factory=list)

    async def next_block(self) -> Sent:
        """The next event or comment the stream writes.

        ``AssertionError`` if the response ends first, which is what a test
        waiting for something that is never coming should be told.
        """
        while not self._blocks:
            assert not self._over, f"the stream ended; it had said {self._read!r}"
            await self._more()
        return self._blocks.pop(0)

    async def upto(self, name: str) -> list[Sent]:
        """Everything up to and including the first event of that AG-UI type."""
        seen = []
        while True:
            block = await self.next_block()
            seen.append(block)
            if block.comment is None and block.type == name:
                return seen

    async def rest(self) -> list[Sent]:
        """Everything the stream writes from here until it ends."""
        while not self._over:
            await self._more()
        kept, self._blocks = self._blocks, []
        return kept

    async def close(self) -> None:
        """Go away in the middle, as a closed tab does.

        ``http.disconnect`` is what a server sends when the client's connection
        is gone, and what Starlette's streaming response listens for.
        """
        self._gone.set()
        await self._served

    async def _more(self) -> None:
        """Wait for the next chunk and put whatever blocks are now whole in reach."""
        chunk = await self._chunks.get()
        if chunk is None:
            self._over = True
            await self._served
            return
        self._read += chunk.decode()
        whole, _, rest = self._read.rpartition(BLOCK)
        if whole:
            self._blocks.extend(parse(whole + BLOCK))
            self._read = rest


@asynccontextmanager
async def streaming(
    app: Any,
    method: str,
    path: str,
    *,
    headers: Mapping[str, str],
    query: str = "",
    body: bytes = b"",
) -> AsyncIterator[Streamed]:
    """Drive that ASGI application for one request, reading it as it is written.

    The scope is the one a server builds; ``scope["app"]`` is Starlette's own
    doing, so the request protection finds the deployment's state where it
    expects to. The request's body is sent in one message, as a small JSON body
    arrives in one.

    Leaving the block closes the connection if the test has not, and waits for
    the application to let go -- so nothing is left running into the next test.
    """
    loop = asyncio.get_running_loop()
    chunks: asyncio.Queue[bytes | None] = asyncio.Queue()
    started: asyncio.Future[tuple[int, dict[str, str]]] = loop.create_future()
    gone = asyncio.Event()
    sent = False

    async def receive() -> dict[str, Any]:
        nonlocal sent
        if not sent:
            sent = True
            return {"type": "http.request", "body": body, "more_body": False}
        await gone.wait()
        return {"type": "http.disconnect"}

    async def send(message: dict[str, Any]) -> None:
        if message["type"] == "http.response.start":
            started.set_result(
                (
                    message["status"],
                    {
                        name.decode("latin-1").lower(): value.decode("latin-1")
                        for name, value in message.get("headers", ())
                    },
                )
            )
        elif message["type"] == "http.response.body":
            await chunks.put(message.get("body", b""))
            if not message.get("more_body", False):
                await chunks.put(None)

    scope = {
        "type": "http",
        "asgi": {"version": "3.0", "spec_version": "2.3"},
        "http_version": "1.1",
        "method": method,
        "scheme": "https",
        "path": path,
        "raw_path": path.encode(),
        "query_string": query.encode(),
        "root_path": "",
        "headers": [
            (name.lower().encode("latin-1"), value.encode("latin-1"))
            for name, value in headers.items()
        ],
        "client": ("127.0.0.1", 1234),
        "server": ("robinauts.example.com", 443),
    }
    served = loop.create_task(app(scope, receive, send))
    try:
        # An application that raised before it answered would otherwise leave
        # this waiting for a response that is never coming; whatever it raised
        # is what a test should be shown.
        await asyncio.wait((started, served), return_when=asyncio.FIRST_COMPLETED)
        if not started.done():
            await served
            raise AssertionError("the application returned without answering")
        status, answered = await started
        yield Streamed(status=status, headers=answered, _chunks=chunks, _served=served, _gone=gone)
    finally:
        gone.set()
        await served

File: frontend/src/chat/assistant-ui/agui/sse.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright The Robinauts Authors

/**
 * Server-sent events, read off a `fetch` body as they arrive.
 *
 * Not `EventSource`: that opens a `GET` of its own, so it cannot carry the
 * body a turn is started with, and it reconnects on its own terms rather than
 * on ours (`docs/specs/wire.md`). What is left of it is this -- the format,
 * which is small -- and the reconnecting, which is `./client.ts`'s, where the
 * position to carry on from is known.
 *
 * The format, as the WHATWG specification has it and as
 * `backend/tests/sse.py` reads it on the other side:
 *
 * - a block is a run of lines ended by a blank one;
 * - a line is `field: value`, with **one** optional space after the colon;
 * - a line beginning with `:` is a comment, which is what the backend's
 *   `: keep-alive` is, and it carries no block of its own;
 * - `data:` may appear more than once in a block and the values are joined
 *   with newlines;
 * - lines end with `\n`, `\r\n` or `\r`, and a chunk may end between the two
 *   halves of a `\r\n`;
 * - a field this reader does not know is ignored, and so is a block that
 *   carried no `data:` at all.
 *
 * `id:` is not the SSE "last event id" state machine of a browser here: this
 * hands each block the id it carried, and the client above decides what the
 * last **real** position was -- the backend numbers only the last wire event
 * derived from each of the run's own events, so a block without an id is
 * never something to re-attach after (`docs/specs/wire.md`).
 */

import { ApiError } from "../../../api/client";

/**
 * How much of one line, and of one block, is read before giving up.
 *
 * **The stream is ours**, and the longest thing the backend ever writes in
 * one block is a message's deltas, which are small. A line or a block past
 * this is a bug of ours or something in front of the deployment writing
 * something else entirely, and either way holding it in a string that grows
 * without bound is how a tab runs out of memory over a connection nobody is
 * watching. Four mebibytes is four times the bound the wire puts on a
 * *request* body (`docs/specs/wire.md`), so nothing legitimate is near it.
 *
 * In UTF-16 units rather than bytes, because that is what a `string`
 * measures; for anything but astral characters the two are the same, and the
 * point of the bound is the order of magnitude.
 */
export const MAX_BLOCK = 4 * 1024 * 1024;

/** What a line or a block past that bound is reported as. */
export const OVERFLOW =
  "this answer's stream sent a block longer than anything we send";

/** One block of a stream that carried data. */
export interface SseBlock {
  /** The `id:` field: a position in the run, or nothing. */
  id: string | null;
  /** The `event:` field, which is the AG-UI type here. */
  event: string | null;
  /** Every `data:` line of the block, joined with newlines. */
  data: string;
}

/**
 * Every data block of that body, in order, until it ends.
 *
 * `signal` is how a caller stops reading: the reader is cancelled, this
 * generator stops, and whatever the signal gave as its reason is thrown --
 * an abort is the caller's own doing and must be told apart from a stream
 * that failed.
 *
 * Whoever stops iterating early -- a `break`, or a `return` -- has the
 * generator's own cleanup cancel the reader, so a stream is never left open
 * behind a caller that has finished with it.
 */
export async function* blocks(
  body: ReadableStream<Uint8Array>,
  signal?: AbortSignal,
): AsyncGenerator<SseBlock> {
  // A caller that has already given up is not given a first block either.
  if (givenUp(signal)) throw signal?.reason;
  const reader = body.getReader();
  // Cancelling makes the pending `read()` resolve as done, which is how a
  // wait for a stream that has gone quiet is ended without a race.
  const stop = () => void reader.cancel().catch(() => undefined);
  signal?.addEventListener("abort", stop, { once: true });
  const decoder = new TextDecoder();
  let buffer = "";
  let pending: string[] = [];
  /** How long the block held so far is, for the bound above. */
  let held = 0;
  // **Where the last search stopped**, carried across chunks, so that a line
  // arriving in a hundred of them is scanned once rather than a hundred
  // times: nothing before this holds a line ending, and the only character
  // that can gain one is the `\r` held back at the very end.
  let from = 0;
  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      // `stream: true`, because a chunk may end in the middle of the bytes
      // of one character and the rest of them is in the next one.
      buffer += decoder.decode(value, { stream: true });
      for (;;) {
        const cut = split(buffer, from);
        if (cut === null) {
          from = buffer.length === 0 ? 0 : buffer.length - 1;
          break;
        }
        buffer = cut.rest;
        from = 0;
        if (cut.line !== "") {
          held += cut.line.length;
          if (held > MAX_BLOCK) throw overflowed();
          pending.push(cut.line);
          continue;
        }
        const block = dispatch(pending);
        pending = [];
        held = 0;
        if (block !== null) yield block;
      }
      if (held + buffer.length > MAX_BLOCK) throw overflowed();
      if (givenUp(signal)) break;
    }
  } finally {
    signal?.removeEventListener("abort", stop);
    // A caller that stopped early, and a body that ended: both leave the
    // reader to be let go of. `cancel` on a finished stream does nothing.
    stop();
  }
  if (givenUp(signal)) throw signal?.reason;
  // A stream that ended without its last blank line still said something.
  // The backend always writes one, and a proxy that cut the connection
  // mid-block is exactly the case a half-written block must not be read as
  // whole -- so what is left is used only when it is a complete block, which
  // here means it parses and carries data.
  const rest = dispatch(pending);
  if (rest !== null) yield rest;
}

/** A line or a block past `MAX_BLOCK`; an `ApiError` like any other failure. */
function overflowed(): ApiError {
  return new ApiError(0, "stream_overflow", OVERFLOW);
}

/**
 * Whether the caller has given up.
 *
 * A function, and not `signal?.aborted` written out three times: a signal is
 * aborted by something outside this loop, and a type checker that has read
 * one test of the flag takes it for settled for the rest of the block.
 */
function givenUp(signal: AbortSignal | undefined): boolean {
  return signal?.aborted === true;
}

/**
 * The first whole line of `buffer` at or after `from`, and what follows it.
 *
 * A `\r` at the very end is held back rather than read as a line ending: the
 * `\n` that would make it one ending rather than two may be the first byte of
 * the next chunk. `from` is where a previous search gave up, which is never
 * past a line ending for exactly that reason.
 */
function split(
  buffer: string,
  from: number,
): { line: string; rest: string } | null {
  for (let at = from; at < buffer.length; at += 1) {
    const character = buffer[at];
    if (character === "\n") {
      return { line: buffer.slice(0, at), rest: buffer.slice(at + 1) };
    }
    if (character === "\r") {
      if (at === buffer.length - 1) break;
      const after = buffer[at + 1] === "\n" ? at + 2 : at + 1;
      return { line: buffer.slice(0, at), rest: buffer.slice(after) };
    }
  }
  return null;
}

/** One block's lines as a block, or nothing where it carried no data. */
function dispatch(lines: readonly string[]): SseBlock | null {
  const data: string[] = [];
  let id: string | null = null;
  let event: string | null = null;
  for (const line of lines) {
    // A comment. The backend's heartbeat is one, and so is anything else a
    // deployment writes to keep a connection open.
    if (line.startsWith(":")) continue;
    const colon = line.indexOf(":");
    const field = colon === -1 ? line : line.slice(0, colon);
    let value = colon === -1 ? "" : line.slice(colon + 1);
    if (value.startsWith(" ")) value = value.slice(1);
    if (field === "data") data.push(value);
    else if (field === "id") id = value;
    else if (field === "event") event = value;
    // Anything else -- `retry:`, a field of a future version -- is ignored,
    // which is what the format asks of a client.
  }
  if (data.length === 0) return null;
  return { id, event, data: data.join("\n") };
}
